### ISSUE TITLE
Upgrade to fflas_ffpack 2.4.3, givaro 4.1.1, and linbox 1.6.3

### ISSUE THREAD

**Issue by d-torrance** [Macaulay2/M2#1114]
These versions were released last summer, and M2 currently doesn't build with them because of a couple small API changes.

I've made the appropriate changes in https://github.com/d-torrance/M2/tree/ffpack-givaro-api-changes

Before I submit a pull request, there's another issue to resolve.  Right now, fflas_ffpack is a submodule, and givaro and linbox are installed from tarballs.  The tarballs are dealt with first, but the `./configure` script for linbox fails because the version of fflas_ffpack is too small (we haven't built it yet!).

So do we:
* Go back to using a tarball to install fflas_ffpack?  (That's what I've done locally to test my patches.)
* Switch to using submodules for givaro and linbox?

**Comment by DanGrayson** [comment]
But linbox is ordinarily not built, so why does it fail?  Are you opting to build it.

If the libraries have to be built before the submodules, then that's an error of mine.  Ideally, we would have targets in M2/GNUmakefile.in such as all-in-linbox and all-in-fflas_ffpack, and then we could include a dependency of the following form:
```
all-in-linbox : all-in-fflas_ffpack
```
in the file GNUmakefile.in.

**Comment by DanGrayson** [comment]
PS: thanks for looking into this!

**Comment by d-torrance** [comment]
Yeah, I was telling `./configure` to build all three so that I could test and make sure everything works.

I'll try adding that target in the `GNUMakefile.in` -- thanks for the suggestion!

**Comment by DanGrayson** [comment]
It won't be so simple to do that, so let me know if you have trouble.

**Comment by d-torrance** [comment]
Yeah, it's tricky!  I got fflas_ffpack to build first by making `all-in-fflas_ffpack` a prerequisite for `all-in-libraries`, but now fflas_ffpack's `configure` fails because it's being built before givaro!

So we need givaro → fflas_ffpack → linbox.  But the outside two are in `libraries` and the middle is in `submodules`.  With the way the Makefile is set up, I don't see a nice way of interrupting `all-in-libraries` to go build fflas_ffpack.

**Comment by DanGrayson** [comment]
Yes, more of the structure of the makefile or even the directory structure needs to be redesigned, so we can express those dependencies.  Since linbox is not normally used, you could submit a pull request, anyway, and things would still work.  Is that right?

Then this issue could stay alive for redesigning the makefile, and I could take it on if you wish.

**Comment by DanGrayson** [comment]
Another issue is the even for the libraries, we don't express the dependencies between them, instead depending on the linear ordering of the list in LIBLIST to do that for us.  If we fix that, we can build the libraries in parallel.

**Comment by d-torrance** [comment]
Yeah, it should work just fine if we just upgrade givaro and fflas_ffpack for now since they're built in the correct order.  I'll submit a pull request soon.

**Comment by DanGrayson** [comment]
Shall I do the rest?

**Comment by d-torrance** [comment]
Yeah -- I wouldn't really know where to start.  :)

**Comment by mikestillman** [comment]
Thanks for doing this.  Also, it seems like the version of givaro we have been using is not compatible with c++17, one thing which we want to try to be able to use.  Hopefully, they have fixed that issue with the latest version.

**Comment by mahrud** [comment]
@DanGrayson by the way, imposing dependencies between the required libraries is something that the CMake build system does well. In that branch I've also made changes to the engine to make it compatible with both the old and newer version of Givaro, and the installed fflas_ffpack version depends on which version is available.

**Comment by mikestillman** [comment]
@mahrud So if I want to use the updated versions in your cmake branch, how do I do that?

**Comment by mahrud** [comment]
With the PR that was just merged, the branches have diverged, so I need to
fix some conflicts first. After that, you can mix the build systems by
building the libraries and programs in CMake, then the rest with configure.

**Comment by d-torrance** [comment]
Fixed in #1115 - closing
